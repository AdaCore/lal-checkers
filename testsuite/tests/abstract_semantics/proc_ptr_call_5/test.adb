procedure Test is

   X : Integer := 42;

   function F(E : Integer) return Integer is
   begin
      return E + 1;
   end F;

   function G(E : Integer) return Integer is
   begin
      return E + X;
   end G;

   P : access function (E : Integer) return Integer;
   B : Boolean;
   R : Integer;
begin
   if B then
      P := F'Access;
   else
      P := G'Access;
   end if;
   R := P(3);
end Test;
