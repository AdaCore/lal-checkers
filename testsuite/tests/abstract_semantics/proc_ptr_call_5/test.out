{
  "always": {
    "assign0": [
      {
        "trace:": [
          "assign0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign1": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign2": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign3": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign4": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assume0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing ()}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign5": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assume1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign6": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign5", 
          "assign6", 
          "assume0", 
          "assume1", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing (), Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign7": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign5", 
          "assign6", 
          "assign7", 
          "assume0", 
          "assume1", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing (), Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign8": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign5", 
          "assign6", 
          "assign7", 
          "assign8", 
          "assume0", 
          "assume1", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing (), Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[4, 45]"
        }
      }
    ], 
    "assume0": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assume0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assume1": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assume1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assume2": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign5", 
          "assume0", 
          "assume1", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing (), Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "split_join0": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign5", 
          "assume0", 
          "assume1", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing (), Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "start0": [
      {
        "trace:": [
          "start0"
        ], 
        "values": {
          "$stack": "({}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ]
  }, 
  "le_t_eq_v": {
    "assign0": [
      {
        "trace:": [
          "assign0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign1": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign2": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign3": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign4": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assume0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing ()}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign5": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assume1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign6": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign6", 
          "assume0", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing ()}", 
          "R": "[-2147483648, 2147483647]"
        }
      }, 
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assign6", 
          "assume1", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign7": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign6", 
          "assign7", 
          "assume0", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing ()}", 
          "R": "[-2147483648, 2147483647]"
        }
      }, 
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assign6", 
          "assign7", 
          "assume1", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign8": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign6", 
          "assign7", 
          "assign8", 
          "assume0", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing ()}", 
          "R": "[4, 4]"
        }
      }, 
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assign6", 
          "assign7", 
          "assign8", 
          "assume1", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[45, 45]"
        }
      }
    ], 
    "assume0": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assume0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assume1": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assume1", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assume2": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assume0", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing ()}", 
          "R": "[-2147483648, 2147483647]"
        }
      }, 
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assume1", 
          "assume2", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "split_join0": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assume0", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{True}", 
          "P": "{Subprogram <SubpBody [\"F\"] 5:4-8:10> capturing ()}", 
          "R": "[-2147483648, 2147483647]"
        }
      }, 
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign5", 
          "assume1", 
          "split_join0", 
          "start0"
        ], 
        "values": {
          "$stack": "({1: '[42, 42]'}, 0)", 
          "B": "{False}", 
          "P": "{Subprogram <SubpBody [\"G\"] 10:4-13:10> capturing (0x00000001)}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "start0": [
      {
        "trace:": [
          "start0"
        ], 
        "values": {
          "$stack": "({}, 0)", 
          "B": "{False, True}", 
          "P": "{[all-path]}", 
          "R": "[-2147483648, 2147483647]"
        }
      }
    ]
  }
}
